lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+
+        # Add leading dimension for time.
+        data.resize(1, len(data))
+        result_cube = Cube(data,
+                           long_name=cube.name(),
+                           dim_coords_and_dims=[(cube.coord('time'), 0),
+                                                (indices, 1)],
+                           aux_coords_and_dims=[(latitude, 1),
+                                                (longitude, 1),
+                                                (utc_offset, 1),
+                                                (bd_ids, 1)],
+                           units=cube.units)
+
+        # Enables use of long_name above for any name, and then moves it
+        # to a standard name if possible.
+        result_cube.rename(cube.name())
+        return result_cube
+
+    def use_nearest(self, cube, sites, neighbours):
+        """
+        Simplest case, in which the diagnostic data value at the nearest grid
+        point, as determined by the chosen PointSelection method, is used for
+        the site.
+
+        Args:
+        -----
+        See process() above.
+
+        """
+        if (not cube.coord_dims(cube.coord(axis='y').name())[0] == 0 or
+                not cube.coord_dims(cube.coord(axis='x').name())[0] == 1):
+            raise Exception("Cube dimensions not as expected.")
+
+        data = cube.data[neighbours['i'], neighbours['j']]
+        return self.make_cube(cube, data, sites)
+
+    def orography_derived_temperature_lapse_rate(self, cube, sites, neighbours,
+                                                 orography, no_neighbours=9):
+        """
+        Crude lapse rate method that uses temperature variation and height
+        variation across local nodes to derive lapse rate. This is highly
+        prone to noise given the small number of points involved and the
+        variable degree to which elevation changes across the small number
+        of points.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            A cube of screen level temperatures at a single time.
+
+        sites/neighbours/no_neighbours : See process() above.
+
+        orography : numpy.array
+            Array of orography data on a grid that corresponds to the grid of
+            the the diagnostic cube.
+
+        Returns:
+        --------
+        Data extracted from the screen level temperature cube at spotdata sites
+        which has then been adjusted using a temperature lapse rate calculated
+        using local variations in temperature with orography.
+
+        """
+        def local_lapse_rate(cube, orography, node_list):
+            """
+            Least-squares fit to local temperature and altitude data for grid
+            points defined by node_list to calculate a local lapse rate.
+
+            """
+            y_data = cube.data[node_list]
+            x_data = orography[node_list]
+            matrix = np.vstack([x_data, np.ones(len(x_data))]).T
```

Comment:
np.vstack should pssoibly be replaced by something else
see: https://docs.scipy.org/doc/numpy/reference/generated/numpy.vstack.html

"This function continues to be supported for backward compatibility, but you should prefer np.concatenate or np.stack. The np.stack function was added in NumPy 1.10."